main.tf

```diff
@@ -1,45 +1,56 @@
 data "aws_sns_topic" "this" {
-  count = "${1 - var.create_sns_topic}"
+  count = "${(1 - var.create_sns_topic) * var.enable}"
 
   name = "${var.sns_topic_name}"
 }
 
 resource "aws_sns_topic" "this" {
-  count = "${var.create_sns_topic}"
+  count = "${var.create_sns_topic * var.enable}"
 
   name = "${var.sns_topic_name}"
 }
 
 locals {
-  sns_topic_arn = "${element(compact(concat(aws_sns_topic.this.*.arn, data.aws_sns_topic.this.*.arn)), 0)}"
+  sns_topic_arn      = "${element(compact(concat(aws_sns_topic.this.*.arn, data.aws_sns_topic.this.*.arn, list(var.enable))), 0)}"
+  function_name_base = "${format("%s-%s", var.lambda_function_name, var.sns_topic_name)}"
```

Comment:
The function name can be passed to the module as an argument if you want to customize it. The function itself does not belong to specific SNS topic, so it can be not bound to just one.